### PR TITLE
Initial tagging system for MooseDocs

### PR DESCRIPTION
**Pull request by cticenhour**
Refs #25387

Still needs some documentation and probably a bit more clean-up, but putting it up for testing and review for those interested.

I would consider this draft of the extension very experimental. It currently replaces data in a user-provided JavaScript file with the combined tag content that it finds when the extension command is triggered. A mechanism to have the filtering page available to the rendered website still needs to be created, so this PR doesn't close the original issue. 

Tagging @GiudGiud for awareness. We can discuss next steps when we're back to work on Tuesday. 

@aeslaughter, as we discussed and if you are interested, feel free to look at and review how @DempseyRogers1994 and I put together the extension.

### CODE REVIEW

File: python/MooseDocs/extensions/tagging.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+This extension defines the tagging command: !tag name=id pairs=key:value. The tag command will accept
+a 'name' parameter string that represents the originating markdown file and is associated with an
+arbitrary list of key:value pairs taken in by the 'pairs' parameter. The tagging extension checks
+that all linked moose pages are unique and will not allow duplicate names in the dictionary.
+Duplicate key value pairs are allowed, but a warning is issued and only one of the pairs is taken in.
+
+ALERT: The tagging extension is experimental! If documentation tagging features are desired, please
+request assistance from the MOOSE Framework development team.
+
+Example Tagger command in *.md:
```
Comment (GiudGiud) [inline python/MooseDocs/extensions/tagging.py:26]:
```suggestion
Example Tag command in *.md:
```

File: python/MooseDocs/extensions/tagging.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        return self._allowed_keys
+
+    def postExecute(self):
+        """
+        At completion execution process, collect and process all current data attributes. Then,
+        collect javascript template content, find/replace, and write to destination.
+        """
+        replace_str = ""
+        for iter in self.getAttributeItems():
+            if bool(re.search('tagger_', iter[0])):
```
Comment (GiudGiud) [inline python/MooseDocs/extensions/tagging.py:95]:
```suggestion
            if bool(re.search('tag_', iter[0])):
```

File: python/MooseDocs/extensions/tagging.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+                msg = "%s: Following 'key' provided more than once; check markdown file: %s"
+                LOG.error(msg, page.name, pair[0])
+            else:
+                good_keys.append([pair[0], pair[1]])
+
+        if len(name) != 0: # Only add to tag database if 'name' is provided
+            page_data = {'name':name, "path":mpath, "key_vals":dict(good_keys)}
+
+            tag_id_name = ''
+            if self.extension.get_tag_data("tagger_" + name):
```
Comment (GiudGiud) [inline python/MooseDocs/extensions/tagging.py:179]:
```suggestion
            if self.extension.get_tag_data("tag_" + name):
```

File: python/MooseDocs/extensions/tagging.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+                msg = "%s: Tag page identifier already exists; not adding the following 'name' to " \
+                    "dictionary: %s"
+                LOG.warning(msg, page.name, name)
+            else:
+                tag_id_name = "tagger_" + name
```
Comment (GiudGiud) [inline python/MooseDocs/extensions/tagging.py:184]:
```suggestion
                tag_id_name = "tag_" + name
```

File: python/MooseDocs/test/extensions/test_tagging.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+    def setupContent(self):
+        config = [dict(root_dir='python/MooseDocs/test/content', content=['js/tagging.js'])]
+        return common.get_content(config, '.md')
+
+    def testStandardUsage(self):
+        with self.assertLogs(level=logging.INFO) as cm: # For warning suppression. TODO: remove cm when non-experimental
+            ast = self.tokenize(self.TEXT)
+        self.assertSize(ast, 1)
+        self.assertEqual(ast(0)['attr_name'], 'tagger_test')
```
Comment (GiudGiud) [inline python/MooseDocs/test/extensions/test_tagging.py:53]:
```suggestion
        self.assertEqual(ast(0)['attr_name'], 'tag_test')
```

File: python/MooseDocs/test/extensions/test_tagging.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        self.assertSize(ast, 0) # No token is created if name is not provided
+
+    def testTaggingDuplicateNamesWarning(self):
+        with self.assertLogs(level=logging.WARNING) as cm:
+            ast1 = self.tokenize(self.TEXT)
+            ast2 = self.tokenize(self.TEXT)
+        self.assertEqual(len(cm.output), 2)
+        self.assertIn('Tag page identifier already exists;', cm.output[1])
+        self.assertSize(ast1, 1)
+        self.assertEqual(ast1(0)['attr_name'], 'tagger_test')
```
Comment (GiudGiud) [inline python/MooseDocs/test/extensions/test_tagging.py:94]:
```suggestion
        self.assertEqual(ast1(0)['attr_name'], 'tag_test')
```

File: python/MooseDocs/extensions/tagging.py [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+
+
+class TaggingCommand(command.CommandComponent):
+    COMMAND= 'tag'
+    SUBCOMMAND= None
+
+    @staticmethod
+    def defaultSettings():
+        settings = command.CommandComponent.defaultSettings()
+        settings['name'] = (None, 'ID name for page and associated key:value category/label pairs.')
```
Comment (GiudGiud) [inline python/MooseDocs/extensions/tagging.py:135]:
```suggestion
        settings['name'] = (None, 'ID name for page and associated key:value category:label pairs.')
```

File: python/MooseDocs/extensions/tagging.py [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+            key_vals=keys.split(':')
+            entry_key_values.append([key_vals[0],key_vals[1]])
+
+        good_keys=[]
+        for pair in entry_key_values:
+            if pair[0] not in self.extension.allowed_keys and len(self.extension.allowed_keys) > 0:
+                msg = "%s: Provided 'key' not in allowed_keys (see config.yml); not adding the " \
+                       "following to the database: %s"
+                LOG.warning(msg, page.name, pair[0])
+            elif len(good_keys) != 0 and pair[0] in good_keys[0]:
```
Comment (GiudGiud) [inline python/MooseDocs/extensions/tagging.py:169]: this only checks the first entry.

I think you want to check all pairs for their first entry
not sure what the one-liner is for that

File: python/doc/content/python/MooseDocs/extensions/tag.md [hunk 9]
```diff
@@ -0,0 +1,54 @@
+# Tagging Extension
+
+The tagging extension provides the "tag" command, which provides a mechanism for setting a set of
+tags made up of arbitrary key:value pairs for categories/labels associated with a given markdown page.
```
Comment (GiudGiud) [inline python/doc/content/python/MooseDocs/extensions/tag.md:4]:
```suggestion
tags made up of arbitrary key:value pairs for categories:labels associated with a given markdown page.
```

File: python/doc/content/python/MooseDocs/extensions/tag.md [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+            - foo
+            - simulation_type
+            - fiscal_year
+```
+
+## Basic Tag
+
+There is only one version of the tag command, demonstrated below. The available settings for the
+tag command are listed in [tagging-settings].
+
```
Comment (GiudGiud) [inline python/doc/content/python/MooseDocs/extensions/tag.md:43]: say what is tagging_one (the name) here

File: python/doc/content/python/MooseDocs/extensions/tag.md [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+during the invocation of the command. The available configuration items for this extension are provided
+in [tagging-config].
+
+!alert! warning title=The tagging extension is experimental!
+This extension is considered to be in an experimental state, as several manual steps are required in
+order to obtain both a set of tagged pages, and a rendered page for filtering against those tags. If
+documentation tagging features are desired, please request assistance from the MOOSE Framework
+development team.
+!alert-end!
+
```
Comment (GiudGiud) [inline python/doc/content/python/MooseDocs/extensions/tag.md:17]: is there a page online for MOOSE that we can link to? that shows this active for a few things?

File: python/MooseDocs/extensions/tagging.py (hunk 8 above)
Comment (cticenhour) [reply python/MooseDocs/extensions/tagging.py:169]: Oh good catch. This was working "well enough" that I didn't catch it in my local testing. I think the fix is to simply perform `pair[0] in good_keys` for that part of the check and Python is smart enough to look through all without extra fuss. Will try that first.

File: python/doc/content/python/MooseDocs/extensions/tag.md (hunk 11 above)
Comment (cticenhour) [reply python/doc/content/python/MooseDocs/extensions/tag.md:17]: There is not with this PR - I could add one to the electromagnetics module (move a couple commits from my other branch to this one) if you want a sample that will be online regularly. Adding it to the test dirs for MooseDocs wouldn't end up being published online.

**Comment by aeslaughter** [comment]
@cticenhour I won't have time to look at this for a little while, so don't wait for me.

File: python/doc/content/python/MooseDocs/extensions/tag.md (hunk 11 above)
Comment (cticenhour) [reply python/doc/content/python/MooseDocs/extensions/tag.md:17]: Once the doco rebuilds, an example page should be live at https://mooseframework.inl.gov/docs/PRs/25388/site/modules/electromagnetics/filter/index.html

File: python/doc/content/python/MooseDocs/extensions/tag.md (hunk 10 above)
Comment (cticenhour) [reply python/doc/content/python/MooseDocs/extensions/tag.md:43]: Done in b2ee9a4

File: python/MooseDocs/extensions/tagging.py (hunk 8 above)
Comment (cticenhour) [reply python/MooseDocs/extensions/tagging.py]: Fixed in e916415

File: python/doc/content/python/MooseDocs/extensions/tag.md (hunk 11 above)
Comment (GiudGiud) [reply python/doc/content/python/MooseDocs/extensions/tag.md:17]: seems it s not working

Comment (cticenhour) [reply python/doc/content/python/MooseDocs/extensions/tag.md:17]: Looks like an issue with the HTML. Browser is looking in `https://mooseframework.inl.gov/js/index-93b559a6.js`, which obviously doesn't exist. [The actual location](https://mooseframework.inl.gov/docs/PRs/25388/site/js/index-93b559a6.js) does however and contains the right information. Will need to explore this tomorrow.

Comment (cticenhour) [reply python/doc/content/python/MooseDocs/extensions/tag.md:17]: Good to go now at the link I posted earlier. Moved to manual relative paths within the HTML, consistent with others that MooseDocs generates.

Comment (GiudGiud) [reply python/doc/content/python/MooseDocs/extensions/tag.md:17]: so the filters arent quite combining the way we wanted.

for example, if you specify benchmark and 2021, we would expect to see less results (only the benchmarks from 2021), not the benchmarks + the models contributed in 2021 (so more results)

Comment (cticenhour) [reply python/doc/content/python/MooseDocs/extensions/tag.md:17]: That would be a @friedmud question. The filtering functionality itself is within his tagview repository and outside of this PR.

Comment (cticenhour) [reply python/doc/content/python/MooseDocs/extensions/tag.md:17]: Also I agree with what you are saying - I would also like to see further filtering with more categories selected.

**Comment by friedmud** [comment]
That can be easily changed.  Just a difference of `||` vs. `&&`

Is that a change you want _right now_?

**Comment by cticenhour** [comment]
@GiudGiud I don't know if a tagview change is necessary for the scope of this PR. I could remove this example, and we can adjust behavior later in the VTB. I will defer to your opinion as reviewer, but wanted to share mine.

**Review by GiudGiud — APPROVED** [review]
looks good. Thanks for taking this over the finish line @cticenhour !

I ll add the tags to the VTB but we can't move forward on the vtb until the change of behavior is made @friedmud

**Comment by cticenhour** [comment]
No problem - thanks for your review and comments!

**Pull request merged**